### PR TITLE
Feature/rx java3

### PR DESCRIPTION
**Pull request by scottslewis**
PR for merge of work by Andreas Larson in 2020...for moving to RxJava3 rather than RxJava2, and work by Scott Lewis adjusting test cases (and in case of fusion removing test cases).

### CODE REVIEW

**Comment by rmichela** [comment]
I'm getting locally reproducible build errors:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:testCompile (default-testCompile) on project reactive-grpc-common: Compilation failure
[ERROR] [snip]/reactive-grpc/common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducerTest.java:[687,6] cannot find symbol
[ERROR]   symbol:   class Test
[ERROR]   location: class com.salesforce.reactivegrpc.common.AbstractSubscriberAndProducerTest
```

**Comment by scottslewis** [comment]
I updated the version number from 1.0.2-SNAPSHOT to 1.2.0-SNAPSHOT and disabled the checkstyle plugin for the benchmarks module (it was trying the checkstyle grpc/protoc-generated code).   It may be possible to change the checkstyle config in a more appropriate way and run checkstyle on the human-created code, but I'm not familiar enough with checkstyle configuration to do that.

**Comment by rmichela** [comment]
Great! It's a lot to take in. +8325 lines. I'll review as fast as I can.

File: rx-java/rxgrpc-stub/pom.xml [hunk 1]
```diff
@@ -28,6 +28,7 @@
         <dependency>
             <groupId>io.grpc</groupId>
             <artifactId>grpc-stub</artifactId>
+            <version>${grpc.version}</version>
```
Comment (rmichela) [inline rx-java/rxgrpc-stub/pom.xml:31]: This gets picked up from the parent pom's dependency management section.

File: common/reactive-grpc-benchmarks/pom.xml [hunk 2]
```diff
@@ -117,6 +117,31 @@
...
+                <executions>
+                    <execution>
+                        <phase>process-sources</phase>
+                        <goals>
+                            <goal>check</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                <!-- Don't run checkstyle for this benchmark project as it seems to be trying to checkstyle
```
Comment (rmichela) [inline common/reactive-grpc-benchmarks/pom.xml:140]: I couldn't reproduce this behavior locally or in CI when running `mvn install` on the top-level pom.

File: rx3-java/rx3grpc/pom.xml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                            <mainClass>com.salesforce.rx3grpc.RxGrpcGenerator</mainClass>
+                        </manifest>
+                    </archive>
+                </configuration>
+            </plugin>
+
+            <plugin>
+                <groupId>com.salesforce.servicelibs</groupId>
+                <artifactId>canteen-maven-plugin</artifactId>
+                <version>1.0.0</version>
```
Comment (rmichela) [inline rx3-java/rx3grpc/pom.xml:72]:
```suggestion
                <version>${canteen.plugin.version}</version>
```

**Comment by scottslewis** [comment]
> 
> 
> I left some comments. Tabs, spaces, and file line endings need some small cleanup.
> 
> I've started a commit that addresses most of the formatting issues. If you add me as a collaborator on your fork, I can push it up to this branch.

I've added you to my fork:  https://github.com/scottslewis/reactive-grpc

File: common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducerTest.java [hunk 4]
```diff
@@ -308,8 +309,8 @@ public void asyncModeWithRacingAndOnErrorTest() {
         Assertions.assertThat(unhandledThrowable).isEmpty();
     }
 
-    @Tag("unstable")
-    @RepeatedTest(2)
+    //@Tag("unstable")
+    //@RepeatedTest(2)
```
[the remark this answers is not in the document]
Comment (scottslewis) [reply common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/AbstractSubscriberAndProducerTest.java:313]: Yes.  For reasons I don't understand this old test was hanging.

**Comment by rmichela** [comment]
My cosmetic changes are in the branch now.

File: common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/StreamObserverAndPublisherRx3Test.java [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+                downstream.request(1);
+            }
+
+            Assertions.assertThat(upstream.requestsQueue)
+                      .hasSize(1)
+                      .containsOnly(DEFAULT_CHUNK_SIZE);
+        }
+    }
+
+    // TODO FIX THIS
```
[the remark this answers is not in the document]
Comment (scottslewis) [reply common/reactive-grpc-common/src/test/java/com/salesforce/reactivegrpc/common/StreamObserverAndPublisherRx3Test.java]: Removed // TODO FIX THIS

File: rx3-java/rx3grpc-stub/BUILD.bazel [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    # Disable error prone build failure (triggered by unused return)
+    javacopts = ["-XepDisableAllChecks"],
+    visibility = ["//visibility:public"],
+    deps = [
+        "//common/reactive-grpc-common",
+        "@com_google_guava_guava",
+        "@io_grpc_grpc_java//context",
+        "@io_grpc_grpc_java//core",
+        "@io_grpc_grpc_java//stub",
+        "@io_reactivex_rxjava2_rxjava",
```
[the remark this answers is not in the document]
Comment (scottslewis) [reply rx3-java/rx3grpc-stub/BUILD.bazel:13]: I'm not familiar with bazel (and apparently Andreas wasn't either) so please fix as needed for this build.

File: pom.xml [hunk 7]
```diff
@@ -63,21 +70,22 @@
         <canteen.plugin.version>1.1.0</canteen.plugin.version>
 
         <!-- Dependency Versions -->
-        <reactive.streams.version>1.0.3</reactive.streams.version>
+        <reactive.streams.version>1.0.2</reactive.streams.version>
```
[the remark this answers is not in the document]
Comment (scottslewis) [reply pom.xml]: Errant search and replace.  I've changed back to 1.0.3

File: rx-java/rxgrpc-stub/pom.xml (hunk 1 above)
Comment (scottslewis) [reply rx-java/rxgrpc-stub/pom.xml]: Added version line.

File: common/reactive-grpc-benchmarks/pom.xml (hunk 2 above)
Comment (scottslewis) [reply common/reactive-grpc-benchmarks/pom.xml]: Not sure why it doesn't work as expected in my dev environment, but if it does in build/yours then feel free to include.

File: rx3-java/rx3grpc/pom.xml (hunk 3 above)
Comment (scottslewis) [reply rx3-java/rx3grpc/pom.xml]: Already fixed...thanks.

**Comment by scottslewis** [comment]
I've made changes for your comments (except for bazel meta-data) and seems to build and test fine on github as per above.  I"m still getting errors in checkstyle for benchmark project using mvn clean install but appears to be just my environment.

**Comment by rmichela** [comment]
Strange. What version of Maven are you running?

**Comment by scottslewis** [comment]
> 
> 
> Strange. What version of Maven are you running?

3.6.3

**Review by rmichela — APPROVED** [review]

**Comment by rmichela** [comment]
LGTM. Thank you _so_ much for seeing this through! 🎉

**Comment by scottslewis** [comment]
Thanks.   Any ETA on release/availability on maven central?

**Comment by rmichela** [comment]
There was a recently reported null pointer exception bug I'd like to fix for 1.2

**Comment by rmichela** [comment]
Released in [1.2.0](https://github.com/salesforce/reactive-grpc/releases/tag/v1.2.0)

**Comment by AndreasLarssons** [comment]
@scottslewis Thanks for fixing my issues and making sure it got merged!!! We changed or internal structure a bit so didn't have time to check back on this. 

Happy to see it got merged!! 🎉 🎉

**Comment by scottslewis** [comment]
@AndreasLarssons my pleasure.  Thanks for doing the original work.   

BTW, @rmichela :  Any chance to get 1.2.1 released soon?   ...i.e. with the fixes for removing use rxjava3 internal packages?

**Comment by rmichela** [comment]
Update released in 1.2.2

**Pull request merged**